src/geometry/euclidean.lean

```diff
@@ -683,4 +683,230 @@ begin
+`orthogonal_projection` and should not be used once that is
+defined. -/
+def orthogonal_projection_fn {s : affine_subspace ℝ V P} (hn : (s : set P).nonempty)
+    (hc : is_complete (s.direction : set V)) (p : P) : P :=
+classical.some $ inter_eq_singleton_of_nonempty_of_is_compl
+  hn
+  (mk'_nonempty p s.direction.orthogonal)
+  ((direction_mk' p s.direction.orthogonal).symm ▸ submodule.is_compl_orthogonal_of_is_complete hc)
+
+/-- The intersection of the subspace and the orthogonal subspace
+through the given point is the `orthogonal_projection_fn` of that
+point onto the subspace.  This lemma is only intended for use in
+setting up the bundled version and should not be used once that is
+defined. -/
+lemma inter_eq_singleton_orthogonal_projection_fn {s : affine_subspace ℝ V P}
+    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) (p : P) :
+  (s : set P) ∩ (mk' p s.direction.orthogonal) = {orthogonal_projection_fn hn hc p} :=
+classical.some_spec $ inter_eq_singleton_of_nonempty_of_is_compl
+  hn
+  (mk'_nonempty p s.direction.orthogonal)
+  ((direction_mk' p s.direction.orthogonal).symm ▸ submodule.is_compl_orthogonal_of_is_complete hc)
+
+/-- The `orthogonal_projection_fn` lies in the given subspace.  This
+lemma is only intended for use in setting up the bundled version and
+should not be used once that is defined. -/
+lemma orthogonal_projection_fn_mem {s : affine_subspace ℝ V P} (hn : (s : set P).nonempty)
+    (hc : is_complete (s.direction : set V)) (p : P) : orthogonal_projection_fn hn hc p ∈ s :=
+begin
+  rw [←mem_coe, ←set.singleton_subset_iff, ←inter_eq_singleton_orthogonal_projection_fn],
+  exact set.inter_subset_left _ _
+end
+
+/-- The `orthogonal_projection_fn` lies in the orthogonal
+subspace.  This lemma is only intended for use in setting up the
+bundled version and should not be used once that is defined. -/
+lemma orthogonal_projection_fn_mem_orthogonal {s : affine_subspace ℝ V P}
+    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) (p : P) :
+  orthogonal_projection_fn hn hc p ∈ mk' p s.direction.orthogonal :=
+begin
+  rw [←mem_coe, ←set.singleton_subset_iff, ←inter_eq_singleton_orthogonal_projection_fn],
+  exact set.inter_subset_right _ _
+end
+
+/-- Subtracting `p` from its `orthogonal_projection_fn` produces a
+result in the orthogonal direction.  This lemma is only intended for
+use in setting up the bundled version and should not be used once that
+is defined. -/
+lemma orthogonal_projection_fn_vsub_mem_direction_orthogonal {s : affine_subspace ℝ V P}
+    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) (p : P) :
+  orthogonal_projection_fn hn hc p -ᵥ p ∈ s.direction.orthogonal :=
+direction_mk' p s.direction.orthogonal ▸
+  vsub_mem_direction (orthogonal_projection_fn_mem_orthogonal hn hc p) (self_mem_mk' _ _)
+
+/-- The orthogonal projection of a point onto a nonempty affine
+subspace, whose direction is complete. -/
```

Comment:
Could you explain (at this point or somewhere else) the connection between the orthogonal projection of an affine subspace and of a complete submodule? I'd like to make it extra clear here because line 751 refers to the "other" `orthogonal_projection`.

Reply:
I've expanded this doc string to discuss that. (I never actually use the other `orthogonal_projection` for anything other than proving this one to be an affine map; all of the properties of this one that I use would work just as well with an unbundled function rather than an `affine_map`. But since it is in fact an affine map, it seems appropriate to define it as an `affine_map` even if I'm never using that property.)